R/legend.grob.R

```diff
@@ -80,154 +80,163 @@ legend.grob <- function(
-						legend.grob.final <- packGrob(
-							frame = legend.grob.final,
-							grob = color.key.grob,
-							row = 3 * (legend.row - 1) + 2,
-							col = 2 * (legend.col - 1) + 1,
-							height = max(unit(legendi[['height']], 'lines'),
-									legend.grob.final$framevp$layout$heights[3 * (legend.row - 1) + 1]),
-							force.height = TRUE,
-							);
-						}
-					else {
-						legend.key <- list(
-							just = c('left', 'top'),
-							between = 0.5,
-							rep = FALSE,
-							rectangles = list(
-								col = legendi[['colours']],
-								size = if (is.null(legendi[['size']])) { size } else { legendi[['size']] },
-								height = 1,
-								border = legendi[['border']]
-								),
-							text = list(
-								legendi[['labels']],
-								cex = label.cex,
-								fontfamily = font.family
-								)
-							);
-
-						# Add the legend to the frame
-						legend.grob.final <- packGrob(
-							frame = legend.grob.final,
-							grob = draw.key(key = legend.key, draw = FALSE),
-							row = 3 * (legend.row - 1) + 2,
-							col = 2 * (legend.col - 1) + 1
-							);
-						}
+			# Figure out where this legend should go in the layout (ordered row-wise)
+			legend.row <- if (i %% layout[1] == 0) { i / layout[1] } else { floor(i / layout[1]) + 1 }
+			legend.col <- if (i %% layout[1] == 0) { layout[1] } else { i %% layout[1] }
+
+			# Create the title of this legend
+			if (!is.null(legendi[['title']])) {
+
+				# Draw a grob representing the title
+				title.x.coord <- 0.5;
+
+				if (title.just == 'left') {
+					title.x.coord <- 0;
 					}
-				);
+				else if (title.just == 'right') {
+					title.x.coord <- 1;
+					}
+
+				title.grob <- textGrob(
+					label = legendi[['title']],
+					just = c(title.just, 'top'),
+					x = title.x.coord,
+					y = 1,
+					gp = gpar(
+						cex = title.cex,
+						fontface = title.fontface,
+						fontfamily = font.family,
+						fill = 'black'
+						)
+					);
+
+				# Get the height of this grob so we can add white space around it
+				title.grob.height <- convertHeight(
+					grobHeight(title.grob),
+					unitTo = 'lines',
+					valueOnly = TRUE
+					);
+
+
+				# Add the title to the frame
+				legend.grob.final <- packGrob(
+					frame = legend.grob.final,
+					grob = title.grob,
+					row = 3 * (legend.row - 1) + 1,
+					col = 2 * (legend.col - 1) + 1,
+					height = max(
+						legend.grob.final$framevp$layout$heights[3 * (legend.row - 1) + 1],
+						unit(title.grob.height + 0.4, 'lines')
+						),
+					force.height = TRUE,
+
+					);
+
+				}
+
+			# Create a key describing the content of the legend
+			# The first column is the coloured rectangles and
+			# the second column is the corresponding text labels
+			if (!is.null(legendi[['continuous']]) && legendi[['continuous']] == TRUE) {
+
+				legendi[['height']] <- if (is.null(legendi[['height']])) { 2 } else { legendi[['height']] };
+				legendi[['width']] <- if (is.null(legendi[['width']])) { 2 } else { legendi[['width']] };
```

Comment:
alighn things like this vertically with padding spaces (e.g. add one space before the <- on line 143